flow/scene_update_context.cc

```diff
@@ -164,55 +178,120 @@ void SceneUpdateContext::CreateFrame(scenic::EntityNode& entity_node,
   }
 
-  bool hit_testable = override_hit_testable.has_value()
-                          ? *override_hit_testable
-                          : view_holder->hit_testable();
-  view_holder->UpdateScene(session_.get(), top_entity_->embedder_node(), offset,
-                           size, SkScalarRoundToInt(alphaf() * 255),
-                           hit_testable);
+  // Set offset.
+  if (view_holder.pending_offset) {
+    view_holder.entity_node.SetTranslation(view_holder.offset.fX,
+                                           view_holder.offset.fY,
+                                           -kScenicZElevationBetweenLayers);
+    view_holder.pending_offset = false;
+  }
 
-  // Assume embedded views are 10 "layers" wide.
-  next_elevation_ += 10 * kScenicZElevationBetweenLayers;
-}
+  // Set HitTestBehavior.
+  if (view_holder.pending_hit_testable) {
+    view_holder.entity_node.SetHitTestBehavior(
+        view_holder.hit_testable
+            ? fuchsia::ui::gfx::HitTestBehavior::kDefault
+            : fuchsia::ui::gfx::HitTestBehavior::kSuppress);
+    view_holder.pending_hit_testable = false;
+  }
 
-void SceneUpdateContext::CreateView(int64_t view_id,
-                                    bool hit_testable,
-                                    bool focusable) {
-  zx_handle_t handle = (zx_handle_t)view_id;
-  flutter::ViewHolder::Create(handle, nullptr,
-                              scenic::ToViewHolderToken(zx::eventpair(handle)),
-                              nullptr);
-  auto* view_holder = ViewHolder::FromId(view_id);
-  FML_DCHECK(view_holder);
-
-  view_holder->set_hit_testable(hit_testable);
-  view_holder->set_focusable(focusable);
-}
+  // Set size and focusable.
+  //
+  // Scenic rejects `SetViewProperties` calls with a zero size.
+  if (view_holder.pending_view_properties && !view_holder.size.isEmpty()) {
+    view_holder.view_holder.SetViewProperties({
+        .bounding_box =
+            {
+                .min = {.x = 0.f, .y = 0.f, .z = -1000.f},
+                .max = {.x = view_holder.size.width(),
+                        .y = view_holder.size.height(),
+                        .z = 0.f},
+            },
+        .inset_from_min = {.x = 0.f, .y = 0.f, .z = 0.f},
+        .inset_from_max = {.x = 0.f, .y = 0.f, .z = 0.f},
+        .focus_change = view_holder.focusable,
+    });
+    view_holder.pending_view_properties = false;
+  }
 
-void SceneUpdateContext::UpdateView(int64_t view_id,
-                                    bool hit_testable,
-                                    bool focusable) {
-  auto* view_holder = ViewHolder::FromId(view_id);
-  FML_DCHECK(view_holder);
+  // Re-attach the ViewHolder at the new place in the scene graph.
+  top_entity_->embedder_node().AddChild(view_holder.opacity_node);
 
-  view_holder->set_hit_testable(hit_testable);
-  view_holder->set_focusable(focusable);
+  // Assume embedded views are 10 "layers" wide.
+  next_elevation_ += 10 * kScenicZElevationBetweenLayers;
```

Comment:
Can 10 * kScenicZElevationBetweenLayers be defined as another constant or collapsed into kScenicZElevationBetweenLayers itself?

Reply:
Will do in next upload